doc/GUIDE.md

```diff
@@ -0,0 +1,910 @@
+has the same layout as the Java `int` type, indicating a 32-bit value, and the
+`long double` type has the same layout as the Java `double` type. (Note that the latter is
+only available on Windows).
+
+### Functions
+
+Let's say we have a main library header file `mylib.h` that contains the following
+function declaration:
+
+```c
+// mylib.h
+
+void foo(int x);
+```
+
+Jextract will generate the following set of methods for this function in the main header
+class it generates:
+
+```java
+// mylib_h.java
+
+public static void foo(int x) { ... } // 1
+
+public static MemorySegment foo$address() { ... } // 2
+public static FunctionDescriptor foo$descriptor() { ... } // 3
+public static MethodHandle foo$handle() { ... } // 4
+```
+
+First and foremost, there is:
+
+1. a static wrapper method that is generated that can be used to
+call the C function.
+
+Besides that, there are also several accessors that return additional meta-data for the
+method:
+
+2. the function's address, represented as a
+  [`MemorySegment`](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/foreign/MemorySegment.html)
+3. the [function descriptor](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/foreign/FunctionDescriptor.html)
+4. the [method handle](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/invoke/MethodHandle.html)
+  returned by the FFM linker, which is used to implement the static wrapper method (1).
+
+The parameter types and return type of this method depend on the
+[carrier types](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/foreign/FunctionDescriptor.html#toMethodType())
+of the layouts that make up the function descriptor of the function, which is itself derived
+from the parsed header files.
+
+### Global Variables
+
+For a global variable declaration in a header file like this:
+
+```c
+// mylib.h
+
+int bar;
+```
+
+Jextract generates the following:
+
+```java
+// mylib_h.java
+
+public static OfInt bar$layout() { ... } // 1
+public static MemorySegment bar$segment() { ... } // 2
+
+public static int bar() { ... } // 3
+public static void bar(int varValue) { ... } // 3
+```
+
+`bar$layout` is the FFM memory layout of the global variable (1), and `bar$segment` is the
+address of the global variable (2).
+
+Besides that, jextract also generates a getter and a setter method to get and set the value
+of the global variable (3). Once again, the parameter and return type of the getter and
+setter depend on the carrier type of the layout of the global variable, which is itself
+derived from the header files.
+
+### Constants (Macros & Enums)
+
+Both macros and enums are translated similarly. If we have a header file containing these
+declarations:
+
+```c
+// mylib.h
+
+#define MY_MACRO 42
+enum MY_ENUM { A, B, C };
+```
+
+Jextract will generate a set of simple getter methods to access the constant values of the
```

Comment:
Nit: maybe omit the word `set` - e.g. just "generate simple getter methods...". Seeing "set" so close to "getter" might be confusing.